spmd/tensor/placement_types.py

```diff
@@ -48,24 +48,19 @@ class _Partial(Placement):
 class DTensorSpec(object):
     mesh: DeviceMesh
     placements: Sequence[Placement]
-    # ndim of the current dist tensor, if only pass in shape, this
-    # field will be updated to len(shape), if not pass in shape, must
-    # pass in ndim
-    ndim: int = -1
     # shape of the current dist tensor, this will be set upon
     # construction of the DTensor, prop rule could read it, and
-    # don't need to set in output spec when calculate the output
+    # would need to set in output spec when calculate the output
```

Comment:
oh is this because we're chaining propagation rules?

Reply:
Yep because we chain the propagation rule, so mm out need to have shape on its DTensorSpec so that we could detect singleton on its generated dtensor spec